yt_dlp/networking/request.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+import copy
+import functools
+import io
+import typing
+from typing import Union, Iterable, Mapping
+try:
+    from urllib.request import _parse_proxy
+except ImportError:
+    _parse_proxy = None
+
+from ..utils import escape_url, sanitize_url, update_url_query, CaseInsensitiveDict
+
+_TYPE_REQ_DATA = Union[bytes, typing.Iterable[bytes], typing.IO, None]
+
+
+class Request:
+    """
+    Represents a request to be made.
+    Partially backwards-compatible with urllib.request.Request.
+
+    @param url: url to send. Will be sanitized.
+    @param data: payload data to send. Must be bytes, iterable of bytes, a file-like object or None
+    @param headers: headers to send.
+    @param proxies: proxy dict mapping of proto:proxy to use for the request and any redirects.
+    @param query: URL query parameters to update the url with.
+    @param method: HTTP method to use. If no method specified, will use POST if payload data is present else GET
+    @param extensions: Dictionary of Request extensions to add, as supported by handlers.
+
+    Apart from the url protocol, proxy dict also supports the following keys:
+    - all: proxy to use for all protocols. Used as a fallback if no proxy is set for a specific protocol.
+    - no: comma seperated list of hostnames (optionally with port) to not use a proxy for.
+    """
+
+    def __init__(
+            self,
+            url: str,
+            data: _TYPE_REQ_DATA = None,
+            headers: typing.Mapping = None,
+            proxies: dict = None,
+            query: dict = None,
+            method: str = None,
+            extensions: dict = None
+    ):
+
+        self._headers = CaseInsensitiveDict()
+        self._data = None
+
+        if query:
+            url = update_url_query(url, query)
+
+        self.url = url
+        self.method = method
+        if headers:
+            self.headers = headers
+        self.data = data  # note: must be done after setting headers
+        self.proxies = proxies or {}
+        self.extensions = extensions or {}
+
+    @property
+    def url(self):
+        return self._url
+
+    @url.setter
+    def url(self, url):
+        self._url = escape_url(sanitize_url(url))
```

Comment:
Add test for this

Reply:
Don't use sanitize_url here, or at least the common_typos part of it (use in YoutubeDL translation layer)